src/surfaceqg.jl

```diff
@@ -0,0 +1,370 @@
+
+  equation = Equation(params, grid)
+
+  return FourierFlows.Problem(equation, stepper, dt, grid, vars, params, dev)
+end
+
+
+# ----------
+# Parameters
+# ----------
+
+"""
+    Params(ν, nν, μ, nμ, calcF!)
+
+Returns the params for Surface QG turbulence.
+"""
+struct Params{T} <: AbstractParams
+       ν :: T         # Buoyancy viscosity
+      nν :: Int       # Buoyancy hyperviscous order
+       μ :: T         # Bottom drag or hypoviscosity
+      nμ :: Int       # Order of hypoviscosity
+  calcF! :: Function  # Function that calculates the buoyancy forcing F
+end
+
+Params(ν, nν) = Params(ν, nν, typeof(ν)(0), 0, nothingfunction)
+
+
+# ---------
+# Equations
+# ---------
+
+"""
+    Equation(params, grid)
+
+Returns the equation for Surface QG turbulence with params p and `grid`.
+"""
+function Equation(params::Params, grid::AbstractGrid)
+  L = @. - params.ν * grid.Krsq^params.nν - params.μ * grid.Krsq^params.nμ
+  CUDA.@allowscalar L[1, 1] = 0
+  return FourierFlows.Equation(L, calcN!, grid)
+end
+
+
+# ----
+# Vars
+# ----
+
+abstract type SurfaceQGVars <: AbstractVars end
+
+struct Vars{Aphys, Atrans, F, P} <: SurfaceQGVars
+        b :: Aphys
+        u :: Aphys
+        v :: Aphys
+       bh :: Atrans
+       uh :: Atrans
+       vh :: Atrans
+       Fh :: F
+  prevsol :: P
+end
+
+const ForcedVars = Vars{<:AbstractArray, <:AbstractArray, <:AbstractArray, Nothing}
+const StochasticForcedVars = Vars{<:AbstractArray, <:AbstractArray, <:AbstractArray, <:AbstractArray}
+
+"""
+    Vars(dev, grid)
+
+Returns the vars for unforced surface QG turbulence on device dev and with `grid`.
+"""
+function Vars(::Dev, grid::AbstractGrid) where Dev
+  T = eltype(grid)
+  @devzeros Dev T (grid.nx, grid.ny) b u v
+  @devzeros Dev Complex{T} (grid.nkr, grid.nl) bh uh vh
+  return Vars(b, u, v, bh, uh, vh, nothing, nothing)
+end
+
+"""
+    ForcedVars(dev, grid)
+
+Returns the vars for forced surface QG turbulence on device dev and with
+`grid`.
+"""
+function ForcedVars(dev::Dev, grid::AbstractGrid) where Dev
+  T = eltype(grid)
+  @devzeros Dev T (grid.nx, grid.ny) b u v
+  @devzeros Dev Complex{T} (grid.nkr, grid.nl) bh uh vh Fh
+  return Vars(b, u, v, bh, uh, vh, Fh, nothing)
+end
+
+"""
+    StochasticForcedVars(dev, grid)
+
+Returns the vars for stochastically forced surface QG turbulence on device
+dev and with grid grid.
+"""
+function StochasticForcedVars(dev::Dev, grid::AbstractGrid) where Dev
+  T = eltype(grid)
+  @devzeros Dev T (grid.nx, grid.ny) b u v
+  @devzeros Dev Complex{T} (grid.nkr, grid.nl) bh uh vh Fh prevsol
+  return Vars(b, u, v, bh, uh, vh, Fh, prevsol)
+end
+
+
+# -------
+# Solvers
+# -------
+
+"""
+    calcN_advection(N, sol, t, clock, vars, params, grid)
+
+Calculates the advection term.
+"""
+function calcN_advection!(N, sol, t, clock, vars, params, grid)
+  @. vars.bh = sol
+  @. vars.uh =   im * grid.l  * sqrt.(grid.invKrsq) * sol
+  @. vars.vh = - im * grid.kr * sqrt.(grid.invKrsq) * sol
```

Comment:
Don't need the extra `.` in `sqrt` --- the `@.` macro in front of this expression will put dots in front of every function. So this line can read

```julia
@. vars.vh = - im * grid.kr * sqrt(grid.invKrsq) * sol
```

The same applies to the line above.

This is not a huge deal of course, since it works as-is.

Reply:
These lines have gone as they were unnecessary (already had the Fourier transformed variables to work with), but I made this mistake a lot through the file so I've corrected all those instances.